File: subarrayEqualsK.py
# Time Complexity : O(n)
# Space Complexity :O(n)
# Did this code successfully run on Leetcode :Yes
# Any problem you faced while coding this :No
# initialize dict , count ad Rsum
# loop over array
# calc Rsum, if sum - k in map then inc count
#  also check if Rsum in map then inc value of Map[Rsum] by 1
# else add Rsum
#return count
def subarraySum(self, nums: List[int], k: int) -> int:
    if len(nums) == 0:
        return 0
    count = 0
    Rsum = 0
    Map = dict()
    Map[0] = 1

    for i in range(len(nums)):
        # x
        Rsum +=nums[i]
        #   x - z = y
        if Rsum - k  in Map:
            count+= Map[Rsum-k]
        if Rsum in Map:
            Map[Rsum]+=1
        else:
            Map[Rsum] = 1
        
    return count

File: SubarrayEqualsK.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach: similar to the contiguous
//subarray problem, we are keeping the running sum here as well. so if we have the sum-k present
//that means that the target is achieved.



import java.util.*;
/**
 * SubarrayEqualsK : SubarrayEqualsK
 *
 * @author : Kavya Mandaliya
 * @version : 1.0 (Nov 04, 2024)
 * @since : 1.0 (Nov 04, 2024)
 */

public class SubarrayEqualsK {
    public int subarraySum(int[] nums, int k) {
        int sum = 0, count = 0;
        Map<Integer, Integer> map = new HashMap<>();
        map.put(0,1);
        for(int i = 0; i < nums.length; i++){
            sum+=nums[i];
            if(map.containsKey(sum-k)){
                count+=map.get(sum-k);
            }
            map.put(sum, map.getOrDefault(sum,0)+1);
        }
        return count;
    }

}
